### ISSUE TITLE
The JIT is incompatible with `static` C functions.

### ISSUE THREAD

**Issue by vext01** [ykjit/yk#647]
The JIT uses extensive use of `dlsym()` to locate the virtual addresses of functions, for example, when outlining a function from a trace. It calls the function, instead of inlining it, which requires knowing the address of the function.

`dlsym()` will only find a function if the symbol has external linkage and appears in the dynamic symbol table.

(Side note: when building an interpreter against yk, `yk-config` provides the `-Wl,--export-dynamic` flag to force symbols that don't need to be exposed to the outside world into the dynamic symbol table so that they can be found. An example of such symbols are those with external linkage, but are being linked into binary executable (as opposed to a shared object). From the linkage POV, there can be no external consumers of the symbols of a binary executable)

C's `static` keyword, when applied to a function, signals that a function is private to its compilation unit. This allows different object files to have distinct functions of the same name, but also requires the symbol to have internal linkage. When a function symbol has internal linkage, `dlsym()` cannot find it.

If you nullify `static` function qualifiers in `clang`:

```diff
diff --git a/clang/lib/CodeGen/CodeGenModule.h b/clang/lib/CodeGen/CodeGenModule.h
index b3354657b237..f906a4cfdce7 100644
--- a/clang/lib/CodeGen/CodeGenModule.h
+++ b/clang/lib/CodeGen/CodeGenModule.h
@@ -1302,7 +1302,7 @@ public:
   llvm::GlobalVariable::LinkageTypes getFunctionLinkage(GlobalDecl GD);
 
   void setFunctionLinkage(GlobalDecl GD, llvm::Function *F) {
-    F->setLinkage(getFunctionLinkage(GD));
+    F->setLinkage(llvm::GlobalValue::ExternalLinkage);
   }
 
   /// Return the appropriate linkage for the vtable, VTT, and type information
```

Then different compilation units are then unable to have functions of the same name without symbol clashes. This manifests as a linker error like:
```
multiple definition of `bool std::__constant_string_p<char>(char const*)'
```

So what do we do? Do we disallow the use of static functions in interpreters (possibly causing the interpreter author to rename functions to avoid symbol clashing?

Should we apply the above clang diff (but put it behind a gate) so that the interpreter author finds out at compile time? Or do we leave clang untouched and let them find out at runtime when the JIT tries to outline a static function, but can't find it?

**Comment by ltratt** [comment]
My inclination is to both name mangle *and* make externally visible. I think that gives us the right C semantics with the yk interoperability we desire, without forcing interpreter authors to make widespread changes.

**Comment by vext01** [comment]
Fixed by #62.
